contracts/ATokenYieldSource.sol

```diff
@@ -0,0 +1,408 @@
+    uint256 _shares = _tokenToShares(_depositAmount);
+
+    require(_shares > 0, "ATokenYS/shares-gt-zero");
+    _supplyToAave(_depositAmount);
+    _mint(_to, _shares);
+
+    emit SuppliedTokenTo(msg.sender, _shares, _depositAmount, _to);
+  }
+
+  /**
+   * @notice Redeems asset tokens from the yield source.
+   * @dev Shares corresponding to the number of tokens withdrawn are burnt from the user's balance.
+   * @dev Asset tokens are withdrawn from Aave, then transferred from the yield source to the user's wallet.
+   * @param _redeemAmount The amount of asset tokens to be redeemed
+   * @return The actual amount of asset tokens that were redeemed.
+   */
+  function redeemToken(uint256 _redeemAmount) external override nonReentrant returns (uint256) {
+    address _underlyingAssetAddress = _tokenAddress();
+    IERC20 _assetToken = IERC20(_underlyingAssetAddress);
+
+    uint256 _shares = _tokenToShares(_redeemAmount);
+    _burn(msg.sender, _shares);
+
+    uint256 _beforeBalance = _assetToken.balanceOf(address(this));
+    _pool().withdraw(_underlyingAssetAddress, _redeemAmount, address(this));
+    uint256 _afterBalance = _assetToken.balanceOf(address(this));
+
+    uint256 _balanceDiff = _afterBalance.sub(_beforeBalance);
+    _assetToken.safeTransfer(msg.sender, _balanceDiff);
+
+    emit RedeemedToken(msg.sender, _shares, _redeemAmount);
+    return _balanceDiff;
+  }
+
+  /**
+   * @notice Claims the accrued rewards for the aToken, accumulating any pending rewards.
+   * @dev Only callable by the owner or manager.
+   * @param _to Address where the claimed rewards will be sent
+   * @return True if operation was successful.
+   */
+  function claimRewards(address _to) external onlyManagerOrOwner returns (bool) {
+    require(_to != address(0), "ATokenYS/payee-not-zero-address");
+
+    address[] memory _assets = new address[](1);
+    _assets[0] = address(aToken);
+
+    (address[] memory _rewardsList, uint256[] memory _claimedAmounts) = rewardsController
+      .claimAllRewards(_assets, _to);
+
+    emit Claimed(msg.sender, _to, _rewardsList, _claimedAmounts);
+    return true;
+  }
+
+  /**
+   * @notice Approve ERC20 tokens other than the aTokens held by this contract to be spent.
+   * @dev This function is only callable by the owner or asset manager.
+   * @dev Allows another contract or address to withdraw funds from the yield source.
+   * @param _token The ERC20 token to approve
+   * @param _spender The spender of the tokens
+   * @param _amount The amount of tokens to approve
+   */
+  function approveERC20(
+    IERC20 _token,
+    address _spender,
+    uint256 _amount
+  ) external onlyManagerOrOwner {
+    require(address(_token) != address(aToken), "ATokenYS/forbid-aToken-approve");
+
+    uint256 _currentAllowance = _token.allowance(address(this), _spender);
+
+    if (_amount > _currentAllowance) {
+      _token.safeIncreaseAllowance(_spender, _amount.sub(_currentAllowance));
```

Comment:
I don't we need this logic as the OZ SafeERC20 adds the amount to the current approved amount.

```
uint256 newAllowance = token.allowance(address(this), spender) + value;
_callOptionalReturn(token, abi.encodeWithSelector(token.approve.selector, spender, newAllowance));
```

And same for the decrease logic.

Reply:
Done: https://github.com/pooltogether/aave-v3-yield-source/pull/1/commits/acc2840ed4d7be0633332d366bbf7dfe7e4b7af1